nipype/pipeline/engine/nodes.py

```diff
@@ -304,124 +312,118 @@ def run(self, updatehash=False):
+        json_unfinished_pat = op.join(outdir, '_0x*_unfinished.json')
+        is_mapnode = isinstance(self, MapNode)
+        need_rerun = (not is_mapnode and
+                      glob(json_pat) and not glob(json_unfinished_pat))
+        if need_rerun:
+            log_debug = config.get('logging', 'workflow_level') == 'DEBUG'
+            logger.debug('[Node] Rerunning "%s"', self.fullname)
+            if log_debug and not hash_exists:  # Lazy logging - only debug
+                exp_hash_paths = glob(json_pat)
+                if len(exp_hash_paths) == 1:
+                    split_out = split_filename(exp_hash_paths[0])
+                    exp_hash_file_base = split_out[1]
+                    exp_hash = exp_hash_file_base[len('_0x'):]
+                    logger.debug("Previous node hash = %s", exp_hash)
+                    try:
+                        prev_inputs = load_json(exp_hash_paths[0])
+                    except:
+                        pass
+                    else:
+                        logging.logdebug_dict_differences(prev_inputs,
+                                                          hashed_inputs)
+            if not force_run and str2bool(self.config['execution']['stop_on_first_rerun']):
+                raise Exception('Cannot rerun when "stop_on_first_rerun" is set to True')
+
+        # Hashfile while running, remove if exists already
+        hashfile_unfinished = op.join(
+            outdir, '_0x%s_unfinished.json' % hashvalue)
+        if op.exists(hashfile):
+            os.remove(hashfile)
+
+        # Delete directory contents if this is not a MapNode or can't resume
+        rm_outdir = not is_mapnode and not (
+            self._interface.can_resume and op.isfile(hashfile_unfinished))
+
+        if rm_outdir:
+            emptydirs(outdir)
+        else:
+            logger.debug(
+                "%s hashfile=%s", '[MapNode] Resume -' if is_mapnode
+                else '[Node] Resume - can_resume=True,', hashfile_unfinished)
+            if is_mapnode:
+                # remove old json files
+                for filename in glob(op.join(outdir, '_0x*.json')):
+                    os.remove(filename)
+
+        # Store runtime-hashfile, pre-execution report, the node and the inputs set.
+        self._save_hashfile(hashfile_unfinished, hashed_inputs)
+        self.write_report(report_type='preexec', cwd=outdir)
+        savepkl(op.join(outdir, '_node.pklz'), self)
+        savepkl(op.join(outdir, '_inputs.pklz'),
+                self.inputs.get_traitsfree())
+        try:
+            self._run_interface(execute=True)
+        except:
+            logger.warning('[Node] Exception "%s" (%s)', self.fullname, outdir)
+            # Tear-up after error
+            os.remove(hashfile_unfinished)
+            os.chdir(cwd)
+            raise
+
+        # Tear-up after success
+        shutil.move(hashfile_unfinished, hashfile)
+        self.write_report(report_type='postexec', cwd=outdir)
```

Comment:
not in this PR, but in a future PR, we can change write_report to take a user defined function.